modules/ROOT/pages/dataweave-type-system.adoc

```diff
@@ -0,0 +1,291 @@
+
+----
+var age: String | Number = ???
+----
+
+A common patter is to use Null with a union type to specify that a type accepts null as a value. In the following example, the variable `a` accepts that the expression `payload.foo` returns `null`.
+
+----
+var a: String| Null = payload.foo
+----
+
+
+=== Intersection Type
+
+The intersection type intersects object types. In this case, the intersection works as the concatenation (`++`) of object types.
+
+_Introduced in DataWeave 2.3.0. Supported by Mule 4.3 and later._
+
+The syntax for intersection type is:
+----
+TypeA & TypeB & ...
+----
+
+In the following example, the intersection concatenates the two object types resulting in the type `{name: String, lastName: String}` This is an open object so it can have more key value pairs. The variable accepts the value assigned to it:
+
+----
+var a: {name: String} & {lastName: String} = {name: "John", lastName: "Smith", age: 34}
+----
+
+In the case of close objects, it returns the concatenation of the object types but resulting in a close object. +
+In the following example, the intersection results in the type `{|name: String, lastName: String|}`. This throws
+an exception because a close object does not accept an object value if there are additional fields in the object (field `age`):
+
+----
+var a: {|name: String|} & {|lastName: String|} = {name: "John", lastName: "Smith", age: 34}
+----
+
+
+=== Literal Types
+
+The literal type represents exactly one value. For example, the String value `"foo"` can be represented with the type `"foo"`.
+
+_Introduced in DataWeave 2.3.0. Supported by Mule 4.3 and later._
+
+The following literal types have been added to the type system:
+
+* String literal types
+* Number literal types
+* Boolean literal types
+
+You can use literal types with union types when you want to declare a type as a finite set of allowed values. For example, the following types declarations are aliases of union and literals types:
+
+----
+type Weekdays = "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday"
+type Days = "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday" | "Saturday" | "Sunday"
+----
+
+In the following example, the variable only accepts the specific literal types (`404` and `500`). The type system ensures that the variable is one of those literal values:
+
+----
+var errorStatusCode: 404 | 500 = payload.statusCode
+----
+
+With function overloading you are able to define different behaviors based on the input argument value:
+
+----
+fun errorHandling(errorCode: 404 | 405, response): String = "Code for error 4XX handling here!"
+fun errorHandling(errorCode: 500 | 501, response): String = "Code for error 5XX handling here!"
+---
+errorHandling(payload.statusCode, payload)
+----
+
+The type system chooses the correct function definition depending on the value of `payload.statusCode`. It is not necessary to check the value of an argument with an if statement to change the behaviour of the function.
```

Comment:
Should we reference the type system as the one making the choice? @TomoDC @machaval

Reply:
This is probably wrong. I didn't notice that the function dispatching works at runtime so the TS didn't participate in the dispatching.
`At runtime, the function dispatching algorithm selects the correct function based on the actual types of the values used` Your correction in the previous comment looks more precise.

Reply:
Updated.